File: backend/src/schemas/events.py
"""Event request/response models."""
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

EventTargetSyncStatusName = Literal["pending", "synced", "failed", "skipped"]
EventScope = Literal["instance", "all_future"]
EventSourceFilter = Literal["fridge", "external", "all"]


class EventTargetView(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    member_id: UUID
    google_event_id: str | None = None
    sync_status: EventTargetSyncStatusName
    retry_count: int
    last_error: str | None = None
    synced_at: datetime | None = None


class EventCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str | None = None
    start_at: datetime
    end_at: datetime
    timezone: str | None = None
    location: str | None = None
    assignee_member_id: UUID | None = None
    car_ids: list[UUID] = Field(default_factory=list)
    rrule: str | None = None


class EventUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    start_at: datetime | None = None
    end_at: datetime | None = None
    timezone: str | None = None
    location: str | None = None
    assignee_member_id: UUID | None = None
    car_ids: list[UUID] | None = None
    rrule: str | None = None


class EventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    family_id: UUID
    title: str
    description: str | None = None
    start_at: datetime
    end_at: datetime
    timezone: str
    location: str | None = None
    assignee_member_id: UUID | None = None
    car_ids: list[UUID]
    rrule: str | None = None
    targets: list[EventTargetView]
    source: Literal["fridge", "external"] = "fridge"
    created_at: datetime
    updated_at: datetime


class ExternalEventResponse(BaseModel):
    """External (read-only) event from a member's Google Calendar.

    Mirrors the shape of EventResponse where it makes sense so the FE can
    render fridge + external events with the same component.
    """
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    family_id: UUID
    member_id: UUID
    google_event_id: str
    title: str | None = None
    description: str | None = None
    start_at: datetime
    end_at: datetime
    location: str | None = None
    is_all_day: bool
    rrule: str | None = None
    source: Literal["external"] = "external"


class EventListResponse(BaseModel):
    fridge: list[EventResponse]
    external: list[ExternalEventResponse]
    total: int

File: frontend/src/lib/api/events.ts
import { http, jsonBody } from "./http";

export type EventSyncStatus = "pending" | "synced" | "failed" | "skipped";

export interface EventTargetView {
  member_id: string;
  google_event_id: string | null;
  sync_status: EventSyncStatus;
  retry_count: number;
  last_error: string | null;
  synced_at: string | null;
}

export type EventSource = "fridge" | "external";

export interface EventResponse {
  id: string;
  family_id: string;
  title: string;
  description: string | null;
  start_at: string;
  end_at: string;
  timezone: string;
  location: string | null;
  assignee_member_id: string | null;
  car_ids: string[];
  rrule: string | null;
  source: EventSource;
  /** Source member id when source=external; null otherwise. */
  source_member_id: string | null;
  targets: EventTargetView[];
  linked_note_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface EventListResponse {
  items: EventResponse[];
  total: number;
}

export interface EventCreateRequest {
  title: string;
  description?: string | null;
  start_at: string; // ISO
  end_at: string;   // ISO
  timezone?: string;
  location?: string | null;
  assignee_member_id?: string | null;
  car_ids?: string[];
  rrule?: string | null;
  linked_note_id?: string | null;
}

export type EventUpdateRequest = Partial<EventCreateRequest>;

export type EventScope = "instance" | "all_future";

export interface EventListFilters {
  from?: string;
  to?: string;
  member_id?: string;
  car_id?: string;
  source?: EventSource | "all";
}

function qs(filters: EventListFilters): string {
  const u = new URLSearchParams();
  if (filters.from) u.set("from", filters.from);
  if (filters.to) u.set("to", filters.to);
  if (filters.member_id) u.set("member_id", filters.member_id);
  if (filters.car_id) u.set("car_id", filters.car_id);
  if (filters.source) u.set("source", filters.source);
  const s = u.toString();
  return s ? `?${s}` : "";
}

export const eventsApi = {
  list: (filters: EventListFilters = {}) =>
    http<EventListResponse>(`/api/events${qs(filters)}`),
  get: (id: string) => http<EventResponse>(`/api/events/${id}`),
  create: (body: EventCreateRequest) =>
    http<EventResponse>("/api/events", {
      method: "POST",
      body: jsonBody(body),
    }),
  update: (id: string, body: EventUpdateRequest, scope: EventScope = "instance") =>
    http<EventResponse>(`/api/events/${id}?scope=${scope}`, {
      method: "PATCH",
      body: jsonBody(body),
    }),
  delete: (id: string, scope: EventScope = "instance") =>
    http<void>(`/api/events/${id}?scope=${scope}`, { method: "DELETE" }),
  resync: (id: string) =>
    http<EventResponse>(`/api/events/${id}/resync`, { method: "POST" }),
};
